hashin.py

```diff
@@ -68,15 +70,37 @@ def run(spec, file, algorithm, verbose=False):
         if verbose:
             _verbose("Latest version for", version)
 
-    # needs to be turned into a list so we know its length
-    hashes = list(get_hashes(data, version, algorithm, verbose=verbose))
+    try:
+        releases = data['releases'][version]
+    except KeyError:
+        raise PackageError('No data found for version {0}'.format(version))
+
+    if python_versions is not None:
+        releases = filter_releases(releases, python_versions)
+
+    if not releases:
+        if python_versions:
+            raise PackageError(
+                "No releases could be found for {0} matching Python versions {1}"
+                .format(spec, python_versions)
+            )
+        else:
+            raise PackageError(
+                "No releases could be found for {0}"
+                .format(spec, python_versions)
+            )
+
+    add_hashes(releases, algorithm, verbose=verbose)
 
     new_lines = ''
     new_lines = '{0}=={1} \\\n'.format(package, version)
     padding = ' ' * 4
-    for i, h in enumerate(hashes):
-        new_lines += '{0}--hash={1}:{2}'.format(padding, algorithm, h)
-        if i != len(hashes) - 1:
+    for i, release in enumerate(releases):
+        new_lines += (
+            '{0}--hash={1}:{2}'
```

Comment:
I really wanted to add comments on this line that included metadata gathered from the URLs, but I couldn't figure out how to get pip to understand comments here.